pipeline/id_minter/src/main/scala/uk/ac/wellcome/platform/idminter/database/IdentifiersDao.scala

```diff
@@ -1,82 +1,134 @@
 package uk.ac.wellcome.platform.idminter.database
 
-import java.sql.SQLIntegrityConstraintViolationException
+import java.sql.{BatchUpdateException, Statement}
 
 import grizzled.slf4j.Logging
 import scalikejdbc._
 import uk.ac.wellcome.models.work.internal.SourceIdentifier
-import uk.ac.wellcome.platform.idminter.exceptions.IdMinterException
 import uk.ac.wellcome.platform.idminter.models.{Identifier, IdentifiersTable}
 
 import scala.concurrent.blocking
 import scala.util.Try
 
+object IdentifiersDao {
+  case class LookupResult(
+    existingIdentifiers: Map[SourceIdentifier, Identifier],
+    unmintedIdentifiers: List[SourceIdentifier])
+  case class InsertResult(succeeded: List[Identifier])
+
+  case class InsertError(failed: List[Identifier],
+                         e: Throwable,
+                         succeeded: List[Identifier])
+      extends Exception
+}
+
 class IdentifiersDao(db: DB, identifiers: IdentifiersTable) extends Logging {
+  import IdentifiersDao._
 
   implicit val session = AutoSession(db.settingsProvider)
 
-  /* An unidentified record from the transformer can give us a list of
-   * identifiers from the source systems, and an ontology type (e.g. "Work").
-   *
-   * This method looks for existing IDs that have matching ontology type and
-   * source identifiers.
-   */
-  def lookupId(
-    sourceIdentifier: SourceIdentifier
-  ): Try[Option[Identifier]] = Try {
-
-    val sourceSystem = sourceIdentifier.identifierType.id
-    val sourceId = sourceIdentifier.value
-
-    // TODO: handle gracefully, don't TryBackoff ad infinitum
-    blocking {
-      debug(s"Matching ($sourceIdentifier)")
-
-      val i = identifiers.i
-      val query = withSQL {
-        select
-          .from(identifiers as i)
-          .where
-          .eq(i.OntologyType, sourceIdentifier.ontologyType)
-          .and
-          .eq(i.SourceSystem, sourceSystem)
-          .and
-          .eq(i.SourceId, sourceId)
-
-      }.map(Identifier(i)).single
-      debug(s"Executing:'${query.statement}'")
-      query.apply()
+  def lookupIds(sourceIdentifiers: Seq[SourceIdentifier]): Try[LookupResult] = {
+    Try {
+      debug(s"Matching ($sourceIdentifiers)")
+      val identifierRows = sourceIdentifiers.map { id =>
+        (id.ontologyType, id.identifierType.id, id.value)
+      }
+      val identifierRowsMap: Map[(String, String, String), SourceIdentifier] =
+        (identifierRows zip sourceIdentifiers).toMap
```

Comment:
Personally I'd be tempted to skip the intermediate variable and the `zip`, but I don't feel strongly:

```suggestion
      val identifierRowsMap: Map[(String, String, String), SourceIdentifier] =
        sourceIdentifiers
          .map { id => (id.ontologyType, id.identifierType.id, id.value) -> id }
          .toMap
```

Reply:
Used 2 variables here as the `identifierRows` is used later